### PR TITLE
Fix bug with no_nncf_trace context

### PR DESCRIPTION
**Pull request by aleksei-kashapov**

### CODE REVIEW

**Comment by aleksei-kashapov** [comment]
@LeonidBeynenson

File: nncf/dynamic_graph/context.py [hunk 1]
```diff
@@ -516,11 +517,16 @@ def reset_graph(self):
...
+    if ctx:
+        if ctx.is_tracing:
+            ctx.disable_tracing()
+            yield
+            if ctx:
+                ctx.enable_tracing()
+        else:
+            yield
+    else:
+        yield
```
Comment (vshampor) [inline nncf/dynamic_graph/context.py:529]: This code looks worse that the one above.
1) Avoid nestings if possible, use compound conditions.
2) Do not do `if a:` when you actually try to test if `a` name points to None. Do `if a is not None`.
3) What problem does this solve? Such explanations should be present in the commit message.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Comment by aleksei-kashapov** [comment]
This PR resolves a problem when we have a nested `no_nncf_trace()` context inside another `no_nncf_trace()`.
Without changes it works as follows: when the program exits the nested context we update a global tracing variable to a disabled tracing state. In that case, another context will be influenced and NNCF starts tracing while it is not expected behavior. 

With this PR nested context checks whether we need to update state from tracing to non_tracing if not we just yield nothing and it doesn't influence another no_nncf_trace() context

File: nncf/dynamic_graph/context.py (hunk 1 above)
Comment (aleksei-kashapov) [reply nncf/dynamic_graph/context.py:525]: 1. Updated
2. Updated
3. Added a description. I hope, now it is more clear

Comment (vshampor) [reply nncf/dynamic_graph/context.py:525]: Please also add a simple test for this logic. Sorry I did not mention it in the previous message.

Comment (aleksei-kashapov) [reply nncf/dynamic_graph/context.py:525]: Add a test to `test_graph_building.py`

**Review by vshampor — APPROVED** [review]

**Pull request merged**